bsdlib/doc/pdn_extension.rst

```diff
@@ -0,0 +1,178 @@
+
+Closing a PDN socket
+********************
+
+.. code-block:: c
+
+   nrf_close(pdn_fd);
+
+The close function for PDN sockets de-registers the associated CID/PDN by deactivating the PDN.
+It also removes the configuration of the CID.
+
+Any socket that has been bound to the APN/PDN must be closed before closing the PDN socket.
+Otherwise, it will not have any transport and will start reporting errors.
+
+Binding a data socket to an APN
+*******************************
+
+.. code-block:: c
+
+   tcp_fd = nrf_socket(NRF_AF_INET, NRF_SOCK_STREAM, NRF_IPPROTO_TCP);
+   nrf_setsockopt(tcp_fd, NRF_SOL_SOCKET, NRF_SO_BINDTODEVICE, "custom.apn", strlen("custom.apn"));
+   nrf_bind(...);
+   # Data will now flow over "custom.apn" instead of Default context.
+   nrf_connect(...);
+   nrf_send(tcp_fd, "data", 4, 0);
+
+The example code shows how to create an IPv4 TCP stream socket and bind it to an APN using the NRF_SO_BINDTODEVICE socket option.
+The APN name must be set up and activated prior to binding it.
+See `Creating a PDN socket`_.
+
+.. note::
+   If the PDN that the socket is being bound to has been deactivated by the network, setting the NRF_SO_BINDTODEVICE option returns an error stating that the APN parameter is not valid.
+   To fix this error, you must create another PDN socket that is associated with the APN, and bind the data socket to it.
+
+Configuring the type of transport
+*********************************
+
+.. code-block:: c
+
+   nrf_setsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_AF, [NRF_AF_INET, NRF_AF_INET6],
+                  sizeof(nrf_sa_family_t) * 2);
+
+You can configure a PDN socket to use a specific Internet family type of transport, for example, IPv4 only, IPv6 only, or dual stack.
+To do so, set the socket option NRF_SO_PDN_AF to request a specific type of connection.
+You must do this after creating the socket and before activating it with ``nrf_connect``.
+
+To retrieve a list of what was activated by the network, call ``getsockopt`` again after activation, using the same socket option.
+
+.. admonition:: Limitation
+
+   The current implementation is hardcoded to always request a dual IP stack.
+
+
+Checking if a PDN is active
+***************************
+
+.. code-block:: c
+
+   nrf_pdn_state_t state;
+   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_STATE, &state, sizeof(nrf_pdn_state_t));
+
+If a data socket is bound to an APN/PDN that is closed, the socket fails with an ``NRF_ENETDOWN`` error.
+
+The example code shows how to check whether a PDN socket is still active.
+However, if there is a long time between transmissions, the PDN socket might appear to still be active, even though the network has deactivated the PDN.
+Only checking the state of the PDN by querying the socket option NRF_SO_PDN_STATE is therefore not reliable.
+Instead, send a transmission on the data socket to trigger the link to become active again.
+The network will then report if the PDN has been deactivated.
```

Comment:
I would rephrase this a bit, perhaps:
```
The example code shows how to check whether a PDN socket is still active.
However, if there is a long time between transmissions, the PDN socket state might be reported as active, even though the network has deactivated the PDN. In this case, sending data on the socket will fail with error NRF_ENETDOWN.
```

Reply:
Updated, but do we then need l. 98 (which says the exact same thing I think)?
And should we add something about that you should check the error code?

Reply:
Yes, those two lines are redundant.

Reply:
OK, removed the first one.